kube-proxy react on Node PodCIDR changes

aojea opened the pull request:
/kind bug

kube-proxy handle node PodCIDR changs
Kube/proxy, in NodeCIDR local detector mode, uses the node.Spec.PodCIDRs
field to build the Services iptables rules.

The Node object depends on the kubelet, but if kube-proxy runs as a
static pods or as a standalone binary, it is not possible to guarantee
that the values obtained at bootsrap are valid, causing traffic outages.

Kube-proxy has to react on node changes to avoid this problems, it
simply restarts if detect that the node PodCIDRs have changed.

In case that the Node has been deleted, kube-proxy will only log an
error and keep working, since it may break graceful shutdowns of the
node.

Fixes #111321, #112739

```release-note
kube-proxy, will restart in case it detects that the Node assigned pod.Spec.PodCIDRs have changed
```

aojea commented:
/sig network
/assign @thockin @danwinship @bowei

danwinship commented:
> PodCIDR is immutable once set but it is possible to delete and
> recreate the Node object to get into this state.

Do we actually expect that to work? (Deleting and recreating a node while kubernetes components are running on that node without restarting them.)

danwinship commented:
It's possible that the node IP would change in that circumstance as well, so this is not the only situation in which kube-proxy would need to restart itself if we want to support that.

bowei commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -645,6 +646,17 @@ func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
+	if proxier.localDetector.IsImplemented() && proxier.localDetector.Mode() == proxyconfigapi.LocalModeNodeCIDR {
+		nodeLocalCIDR := proxier.localDetector.IfLocal()[1]
+		if !utilproxy.NodeContainsPodCIDR(node, nodeLocalCIDR) {
+			klog.ErrorS(nil, "Inconsisten statu, kube-proxy configed Local CIDR not present on Node.Spec.PodCIDRs", "node", klog.KObj(node), "local CIDR", nodeLocalCIDR, "PodCIDRs", node.Spec.PodCIDRs)
```
spelling?

Make this log message more direct:

"Current Node PodCIDR(s) not longer matches the PodCIDR kube-proxy saw at startup, restarting"

bowei commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -966,6 +967,17 @@ func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
+	if proxier.localDetector.IsImplemented() && proxier.localDetector.Mode() == proxyconfigapi.LocalModeNodeCIDR {
+		nodeLocalCIDR := proxier.localDetector.IfLocal()[1]
+		if !utilproxy.NodeContainsPodCIDR(node, nodeLocalCIDR) {
+			klog.ErrorS(nil, "Inconsisten statu, kube-proxy configed Local CIDR not present on Node.Spec.PodCIDRs", "node", klog.KObj(node), "local CIDR", nodeLocalCIDR, "PodCIDRs", node.Spec.PodCIDRs)
```
Match this log message with the one above.

Also, this code looks to be the same, we might want to factor it out to a func? (name it MaybeRestartOnInconsistentPodCIDR()?)

aojea replied:
I thought about that, the reasons I didn't do it is because these are 7 lines of code with a lot of deps on `proxier` methods ... the refactor to a func is not going to save more lines and will make it harder to read IMHO ...

aojea replied:
to be honest, I like your idea, but with current state of the code on both proxies, I don't know if it is better to keep them isolated or to try to consolidate them

aojea commented:
> It's possible that the node IP would change in that circumstance as well, so this is not the only situation in which kube-proxy would need to restart itself if we want to support that.

one bug at a time ;) ... until now we only have evidence of this one

bowei commented on pkg/proxy/util/utils.go:
```diff
@@ -587,3 +587,16 @@ func RevertPorts(replacementPortsMap, originalPortsMap map[netutils.LocalPort]ne
 	}
 }
+
+// NodeContainsPodCIDR returns true if the node contains the passed cidr
```
minor: Containes => Has (contains with CIDR indicates containment, e.g. 10.0.0.0/24 is contained inside 10.0.0.0/16)

aojea replied:
yep

bowei commented:
Left a comment about potentially being able to restrict the surface area of the change.

bowei reviewed (commented):
Sorry, forgot to hit send on the review

aojea commented on cmd/kube-proxy/app/server_others_test.go:
```diff
@@ -510,25 +532,38 @@ func Test_getLocalDetector(t *testing.T) {
+			r, err := getLocalDetector(c.mode, c.config, c.ipt, c.nodeInfo)
+			if c.errExpected {
+				if err == nil {
+					t.Fatalf("Case[%d] Expected error, but succeeded with %v", i, r)
```
yeah, we don't, is redundant

aojea commented on cmd/kube-proxy/app/server_others_test.go:
```diff
@@ -510,25 +532,38 @@ func Test_getLocalDetector(t *testing.T) {
-			t.Errorf("Case[%d] Unexpected detect-local implementation, expected: %q, got: %q", i, c.expected, r)
-		}
+			if err != nil {
+				t.Errorf("Case[%d] Error resolving detect-local: %v", i, err)
```
yes

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -639,6 +640,17 @@ func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
 		return
 	}
 
+	// kube-proxy in LocalModeNodeCIDR mode may cache stale Node.PodCIDR.
```
I've added a new method to return the Value() of the local traffic discriminator

bowei replied:
I meant instead of exposing the Value() and putting the logic outside, we expose just the notion that the detector is still valid (which is what is being checked here).

so the proxier call site just has

```golang
if proxier.localDetector.IsImplemented() {
  if err := proxier.localDetector.IsValid(node); err != nil {
    klog.ErrorS(...)
    klog.FlushAndExit(...)
  }
}
```

aojea replied:
yeah,  but localDetectors can use CIDRs or interfaces names/prefixes, or whatever we want to implement in the feature, adding a new method `IsValid(node *v1.Node)` is only valid for one of the traffic detector modes.
Do we want to add such specific method to an `Interface`?

bowei replied:
I'm suggesting removing Mode() and Value() and instead have a IsValid(node *v1.Node).

For detectors that are always Valid, this can always return `true`.

aojea replied:
ah, that makes more sense

danwinship commented:
/hold

> > It's possible that the node IP would change in that circumstance as well, so this is not the only situation in which kube-proxy would need to restart itself if we want to support that.
> 
> one bug at a time ;) ... until now we only have evidence of this one

No, I'm disagreeing that there is evidence of a bug. I see a user saying "if I intentionally break my node by doing unsupported things, it stops working", and the correct response to that is "don't do that then".

bowei commented:
This is a behavior that is possible -- and for full disclosure -- preemptible VMs can trigger this path. I would disagree that there is no bug -- the Node spec has changed but the kube-proxy stays on the stale information.

aojea commented:
it's a fair question, let me try to create a reproducer ... it has to be after code freeze though

danwinship commented:
> This is a behavior that is possible -- and for full disclosure -- preemptible VMs can trigger this path. I would disagree that there is no bug -- the Node spec has changed but the kube-proxy stays on the stale information.

Clearly there's a bug _somewhere_. But it's not clear to me that the bug is "kube-proxy needs to monitor an immutable field in case it mutates". I'd lean more towards "GCP's preemptible VMs feature needs to do a better job of cleaning up stale node state after deleting a `Node`". There may be other components that also get confused in this case. (Most obviously, the network plugin is also likely to be caching the `podCIDR[s]` value, but there may be other components caching other Node fields/annotations that are normally expected to be immutable too.)

> let me try to create a reproducer

If this is a supported behavior then we should have an e2e test

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -639,6 +640,17 @@ func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
+	if proxier.localDetector.IsImplemented() && proxier.localDetector.Mode() == proxyconfigapi.LocalModeNodeCIDR {
+		nodeLocalCIDR := proxier.localDetector.IfLocal()[1]
+		if !utilproxy.NodeContainsPodCIDR(node, nodeLocalCIDR) {
+			klog.ErrorS(nil, "Current Node PodCIDR(s) not longer matches the PodCIDR kube-proxy saw at startup, restarting", "node", klog.KObj(node), "local CIDR", nodeLocalCIDR, "PodCIDRs", node.Spec.PodCIDRs)
```
This restarts if _any_ node's PodCIDR doesn't match, doesn't it? You need to check `node.Name` or something...

But also, rather than checking the PodCIDR specifically, kube-proxy could just restart if its own Node gets deleted, or if its `UID` or `creationTimestamp` changes. (I think the shared informer code will potentially squash a delete+create into an update like that?)

aojea replied:
this happens at startup
 
https://github.com/kubernetes/kubernetes/blob/70dcb0f1297e0a7b183720cfacece73c57ee1afc/cmd/kube-proxy/app/server_others.go#L166-L173

aojea replied:
hold on, your into something, that is not the informer code, that is the code that checks for the PodCIDR on the nodeName before start.

The informer runs only on the node name

https://github.com/kubernetes/kubernetes/blob/d96af5f2761c23eddde206aee7b7147a9692b737/cmd/kube-proxy/app/server.go#L753-L766


.... but only works with TopologyAwareHints feature gate enabled @bowei

danwinship commented:
(:+1: on using `t.Run`)

aojea commented on pkg/proxy/topology.go:
```diff
@@ -145,6 +145,10 @@ func CategorizeEndpoints(endpoints []Endpoint, svcInfo ServicePort, nodeLabels m
 // * All of the endpoints for this Service have a topology hint
 // * At least one endpoint for this Service is hinted for this node's zone.
 func canUseTopology(endpoints []Endpoint, svcInfo ServicePort, nodeLabels map[string]string) bool {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.TopologyAwareHints) {
```
this was documented in the function comment but missing in the code, added an unit test too

aojea commented:
/hold cancel

squashed some commits so it can be backported

bowei commented:
Left one comment to reduce exposed surface area. otherwise looks good to me
...

bowei commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -643,6 +644,17 @@ func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
 
+	// kube-proxy in LocalModeNodeCIDR mode may cache stale Node.PodCIDR.
+	// It should restart on those situations to reconcile the state.
+	// Ref: https://issues.k8s.io/111321
```
Do we know for sure that a kp restart on UID does not have other unintended consequences? I do like the big hammer though if we have a good sense that this won't cause problems when backporting.

danwinship replied:
Restarting kube-proxy at any time ought to be totally safe. It will pause updates for however long it takes to restart and resync the caches, but other than that it should be a no-op.

bowei replied:
+1

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -615,19 +617,21 @@ func (proxier *Proxier) OnNodeAdd(node *v1.Node) {
-	if reflect.DeepEqual(proxier.nodeLabels, node.Labels) {
-		return
-	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.TopologyAwareHints) {
+		if reflect.DeepEqual(proxier.nodeLabels, node.Labels) {
+			return
```
sounds much better with your suggestions

aojea commented:
retaking this, sorry for the delay, I've been busy with other projects :/

aojea commented:
all yours @danwinship @thockin @bowei 

@liggitt we have to detect when the Node has changed, it starts an informer during bootstrap filtering by the metadata.name ... is it safe to use the Add() event handler to get the node.UID so we can compare later it has changed?
or should I get the UID during bootstrap?

liggitt commented:
> @liggitt we have to detect when the Node has changed, it starts an informer during bootstrap filtering by the metadata.name ... is it safe to use the Add() event handler to get the node.UID so we can compare later it has changed?

the Add handler will be called on first observation of the watched object... you could also initialize the seen uid in your sync loop if uninitialized

khenidak commented:
While the code is /lgtm i think we have a good opportunity to remove the duplication across proxies by an embedded struct or something like that. if you don't have the time then let us leave a note and i will remove the duplication in code later on.

aojea commented:
> While the code is /lgtm i think we have a good opportunity to remove the duplication across proxies by an embedded struct or something like that. if you don't have the time then let us leave a note and i will remove the duplication in code later on.

there is some intention to backport, so I want to minimize the code change, a refactor like that will make it harder to justify

aojea commented:
are we still interested in this?

bowei commented:
Yes, what are the blockers  -- this for sure is a bug.

aojea commented:
@thockin @danwinship are we ok with this approach?

bobbypage commented:
/cc

aojea commented:
/hold

this is not a problem if kube-proxy follows the node lifecycle, let me verify some things before

aojea commented:
I think that this is not the right solution, will update it later

aojea commented:
> I think that this is not the right solution, will update it later

Talked with @liggitt offline, this is the only solution, considering that kube-proxy can run in standalone or as a static pod, hence it can start before the kubelet and, if the node has the same nodeName, it will read the old object.

/hold cancel

aojea commented:
/assign @liggitt

bowei commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -581,16 +584,36 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
+	proxier.mu.Lock()
+	nodeUID := string(node.UID)
+	if proxier.nodeUID == "" {
+		proxier.nodeUID = nodeUID
```
we should log info here that we are setting to the UID. 

klog.InfoS("Setting current Node UID", "nodeUID", nodeUID)

bowei commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -581,16 +584,36 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
 // OnNodeAdd is called whenever creation of new node object
 // is observed.
 func (proxier *Proxier) OnNodeAdd(node *v1.Node) {
+	// This can not happen, the informer is filtering the node by name
```
Instead of a comment, you want to remove this code in another PR?

aojea replied:
I've created a separate handler to handle the node events, it is much cleaner and apply to both proxiers ... it also allows to grow independently of the impementation

bowei commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -606,8 +629,24 @@ func (proxier *Proxier) OnNodeAdd(node *v1.Node) {
 // node object is observed.
 func (proxier *Proxier) OnNodeUpdate(oldNode, node *v1.Node) {
 	if node.Name != proxier.hostname {
+		// This can not happen, the informer is filtering the node by name
```
same here...

aojea committed:
kube-proxy: gate topology correctly

liggitt commented:
/unassign

I'll defer review of this one to network folks

bowei commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -606,8 +629,24 @@ func (proxier *Proxier) OnNodeAdd(node *v1.Node) {
+		klog.FlushAndExit(klog.ExitFlushTimeout, 1)
+	}
+
+	// Reboot if the Node has changed, same Node name but different UID.
```
suggestion: take the code that is repeated and put it in  

```
func (proxier* Proxier) assertNodeValidOrExit(nodeUID string) {

}
```

aojea committed:
kube-proxy handle node PodCIDR changs

Kube/proxy, in NodeCIDR local detector mode, uses the node.Spec.PodCIDRs
field to build the Services iptables rules.

The Node object depends on the kubelet, but if kube-proxy runs as a
static pods or as a standalone binary, it is not possible to guarantee
that the values obtained at bootsrap are valid, causing traffic outages.

Kube-proxy has to react on node changes to avoid this problems, it
simply restarts if detect that the node PodCIDRs have changed.

In case that the Node has been deleted, kube-proxy will only log an
error and keep working, since it may break graceful shutdowns of the
node.

aojea committed:
kube-proxy wait for cluster cidr skip delete events

aojea commented:
/hold cancel

this is much cleaner

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -760,20 +759,20 @@ func (s *ProxyServer) Run() error {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", s.NodeRef.Name).String()
+		}))
+	nodeConfig := config.NewNodeConfig(currentNodeInformerFactory.Core().V1().Nodes(), s.ConfigSyncPeriod)
+	nodeConfig.RegisterEventHandler(&proxy.NodeStateHandler{})
```
this appends the eventHandlers

dcbw commented:
Implementation LGTM. I'll leave the larger/philsophical questions about Node object changes to others...

aojea commented:
> 

I've reduced the scope to PodCIDRs changes only when using nodeCIDR local detector, this is much safer now

aojea commented on cmd/kube-proxy/app/server.go:
```diff
@@ -760,20 +760,23 @@ func (s *ProxyServer) Run() error {
+	nodeConfig := config.NewNodeConfig(currentNodeInformerFactory.Core().V1().Nodes(), s.ConfigSyncPeriod)
+	// https://issues.k8s.io/111321
+	if s.localDetectorMode == kubeproxyconfig.LocalModeNodeCIDR {
+		nodeConfig.RegisterEventHandler(&proxy.NodePodCIDRHandler{})
```
check I changed the logic to panic on PodCIDR only and added test cases to cover the different states, if no PodCIDR exists in the handler it means that the handler wasn't initialized, os it uses the new value for that

thockin reviewed (commented):
Thanks!

/lgtm
/approve

aojea commented:
I had to rebase @thockin :/

thockin commented:
Thanks!

/lgtm
/approve
